src/test/java/org/stellar/sdk/Sep10ChallengeTest.java

```diff
@@ -968,6 +959,98 @@ public void testReadChallengeTransactionInvalidAdditionalOpsOfOtherTypes() throw
+    long end = now + 300;
+    TimeBounds timeBounds = new TimeBounds(now, end);
+
+    Transaction transaction = null;
+    try {
+      transaction = Sep10Challenge.newChallenge(
+        server,
+        network,
+        client.getAccountId(),
+        domainName,
+        timeBounds
+      );
+    } catch (InvalidSep10ChallengeException e) {
+      fail("Should not have thrown any exception.");
+    }
+
+    Sep10Challenge.ChallengeTransaction challengeTransaction = Sep10Challenge.readChallengeTransaction(transaction.toEnvelopeXdrBase64(), server.getAccountId(), Network.TESTNET, new String[]{"example3.com", "example2.com", "example.com"});
+    assertEquals(new Sep10Challenge.ChallengeTransaction(transaction, client.getAccountId(), domainName), challengeTransaction);
+  }
+
+  @Test
+  public void testReadChallengeTransactionInvalidDomainNamesMismatch() throws IOException {
+    KeyPair server = KeyPair.random();
+    KeyPair client = KeyPair.random();
+    Network network = Network.TESTNET;
+    String domainName = "example.com";
+
+    long now = System.currentTimeMillis() / 1000L;
+    long end = now + 300;
+    TimeBounds timeBounds = new TimeBounds(now, end);
+
+    Transaction transaction = null;
+    try {
+      transaction = Sep10Challenge.newChallenge(
+        server,
+        network,
+        client.getAccountId(),
+        domainName,
+        timeBounds
+      );
+    } catch (InvalidSep10ChallengeException e) {
+      fail("Should not have thrown any exception.");
+    }
+
+    try {
+      Sep10Challenge.readChallengeTransaction(transaction.toEnvelopeXdrBase64(), server.getAccountId(), Network.TESTNET, new String[]{"example2.com", "example1.com"});
+      fail();
+    } catch (InvalidSep10ChallengeException e) {
+      assertEquals("The transaction's operation key name does not include one of the expected home domains.", e.getMessage());
+    }
+  }
+
+  @Test
+  public void testReadChallengeTransactionInvalidDomainNamesEmpty() throws IOException {
+    KeyPair server = KeyPair.random();
+    KeyPair client = KeyPair.random();
+    Network network = Network.TESTNET;
+    String domainName = "example.com";
+
+    long now = System.currentTimeMillis() / 1000L;
+    long end = now + 300;
+    TimeBounds timeBounds = new TimeBounds(now, end);
+
+    Transaction transaction = null;
+    try {
+      transaction = Sep10Challenge.newChallenge(
+        server,
+        network,
+        client.getAccountId(),
+        domainName,
+        timeBounds
+      );
+    } catch (InvalidSep10ChallengeException e) {
+      fail("Should not have thrown any exception.");
+    }
+
+    try {
+      Sep10Challenge.readChallengeTransaction(transaction.toEnvelopeXdrBase64(), server.getAccountId(), Network.TESTNET, new String[]{});
+      fail();
+    } catch (InvalidSep10ChallengeException e) {
+      assertEquals("The transaction's operation key name does not include one of the expected home domains.", e.getMessage());
```

Comment:
❗ I think this exception should actually be an `InvalidArgumentException`. The issue is the calling code didn't provide a domain name in the list, and not that the challenge is missing any specific data.

Reply:
I think the one exception is if the list is empty like it is in this test. But if a non-empty list is passed, it should be an `InvalidSep10ChallengeException`.

Reply:
Hi, I added additional verification for it and thrown an `IllegalArgumentException` exception.

According to the discussion, should `IllegalArgumentException` exception be thrown in the code?
https://github.com/stellar/java-stellar-sdk/blob/7b3d60e37c/src/main/java/org/stellar/sdk/Sep10Challenge.java#L86

Reply:
Yes probably, but let's not change that in the pr since it's unrelated and would be a breaking change to fix. We should fix it in another pr and consider how to handle the breaking change. It might be fine if the SDKs major version is still 0.